proposals/low-level-struct-improvements.md

```diff
@@ -0,0 +1,972 @@
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            if (index >= _list.Count)
+            {
+                return ref _default;
+            }
+
+            return ref _list[index];
+        }
+    }
+}
+```
+
+Members which contain the `[RefEscapes]` attribute cannot be used to implement 
+interface members. This would hide the lifetime nature of the member at 
+the `interface` call site and would lead to incorrect lifetime calculations.
+
+To account for this change the "Parameters" section of the span safety document
+will be updated to include the following:
+
+- If the parameter is the `this` parameter of a `struct` type, it is 
+*ref-safe-to-escape* to the top scope of the enclosing method unless the 
+method is annotated with `[RefEscapes]` in which case it is *ref-safe-to-escape*
+outside the enclosing method.
+
+Misc Notes:
+- A member marked as `[RefEscapes]` can not implement an `interface` method.
+- The `RefEscapesAttribute` will be defined in the 
+`System.Runtime.CompilerServices` namespace.
+
+### Safe fixed size buffers
+The language will relax the restrictions on fixed sized arrays such that the 
+can be declared in safe code and the element type can be managed or unmanaged. 
+This will make types like the following legal:
+
+```cs
+internal struct CharBuffer
+{
+    internal fixed char Data[128];
+}
+```
+
+These declarations, much like their `unsafe` counter parts, will define a 
+sequence of `N` elements in the containing type. These members can be accessed 
+with an indexer and can also be converted to `Span<T>` and `ReadOnlySpan<T>` 
+instances.
+
+When indexing into a `fixed` buffer of type `T` the `readonly` state of the 
+container must be taken into account.  If the container is `readonly` then the
+indexer returns `ref readonly T` else it returns `ref T`. 
+
+Accessing a `fixed` buffer without an indexer has no natural type however it is
+convertible to `Span<T>` types. In the case the container is `readonly` the 
+buffer is implicitly convertible to `ReadOnlySpan<T>`, else it can implicitly 
+convert to `Span<T>` or `ReadOnlySpan<T>` (the `Span<T>` conversion is 
+considered *better*). 
+
+The resulting `Span<T>` instance will have a length equal to the size declared
+on the `fixed` buffer. The *safe-to-escape* scope of the returned value will
+be equal to the *safe-to-escape* scope of the container.
+
+For each `fixed` declaration in a type where the element type is `T` the 
+language will generate a corresponding `get` only indexer method whose return
+type is `ref T`. The indexer will be annotated with the `[RefEscapes]` attribute
+as the implementation will be returning fields of the declaring type. The 
+accessibility of the member will match the accessibility on the `fixed` field.
+
+For example, the signature of the indexer for `CharBuffer.Data` will be the 
+following:
+
+```cs
+[RefEscapes]
+internal ref char <>DataIndexer(int index) => ...;
```

Comment:
I would suggest also adding ordinary byval getter and setter and select which one to use based on the same rules as with ordinary arrays. (Arrays implicitly have 3 index accessors - get, set and address).

The reasons for why this would be convenient are:
- evaluation order of arguments in indexed assignments will match ordinary arrays and common collections. I.E. you will get `IndexOutOfRange` after evaluating RHS, for example. 
It is a small thing but consistency could be nice when switching between inline arrays and ordinary arrays.
- byval accessors may compose better with features like `await`. 
Although I think the byref accessor in this case could be spillable too, there is some simplicity when spilling values. Even for ordinary arrays it is a bit simpler.

Reply:
This is a good point. Another benefit of going with the three accessor pattern is that it will benefit language interop. Think these could be structured as normal named indexers that *should* just work with existing languages.

Reply:
Would this mean users could define:
```csharp
public int X
{
    get => ...
    set => ...
    ref => ...
}
```
?

So far the framework has "worked around" this by defining `Value` and `ValueRef` versions of the property. But the above would allow you to extend an existing property and to light up as possible, which would be amazing

Reply:
> Would this mean users could define

No, this would be an implementation detail of safe `fixed` buffers. The idea being make them as array-like as possible.